feat(api): Added upload transcript api

cheblankenshipUTD opened the pull request:
## Overview

Issue 123

[URL]
https://github.com/UTDNebula/planner/issues/123

This task relates to issue 123 'parse credit history from transcript'.
I added the upload.ts file which handles the parsing when a user uploads the transcript.

upload.ts is added to directory "planner/api/transcript/"

## Other Notes

Needs discussion about where to locate the UI component (upload transcript button).

Commit:
feat(api): Added upload transcript api

ZzRanger reviewed (commented):
Haha sorry I mispoke in DMs move the file to /pages/api/transcript and rename the file to index.ts

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+import path from 'path';
+import pdf from 'pdf-parse';
+
+const courseCode = [
```
Move the courseCode data into a json file called "courseCode.json" under /data

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    /* Move uploaded files to directory */
+    for (const file of files) {
```
If I'm not mistaken, the transcript that the user uploads is not the one that gets parsed. Can you rewrite so that the transcript uploaded is the one getting parsed?

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+      const takenCourses = Array.from(new Set(temp));
+
+      // Courses
+      for (let k = 0; k < takenCourses.length; k++) {
```
Instead of console logging can we have the API return this as a json response?

cheblankenshipUTD replied:
I with res.json instead of console logging and tested the JSON response on Chrome > Network log.

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+import formidable, { File } from 'formidable';
```
Add this and pdf-parse to your package.json then push that 

command is ```npm i pdf-parse formidable```

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+
+    /* Parse the uploaded file */
+    const dataBuffer = fs.readFileSync(
+      '/Users/cheblankenship/utd/z_other/nebula_labs/local_personal_code/transcript.pdf',
```
Related to comment above

cheblankenshipUTD replied:
The hard-coded file path I accidentally set is now fixed and it should dynamically get the file path.

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+    });
+    form.on('end', () => resolve(files));
+    form.on('error', (err) => reject(err));
+    form.parse(req, () => {
```
Is this logic parsing the files uploaded and if so how?

ZzRanger commented on api/transcript/upload.ts:
```diff
@@ -0,0 +1,202 @@
+
+const handler = async (req: NextApiRequest, res: NextApiResponse) => {
+  let status = 200,
+    resultBody = { status: 'ok', message: 'Files were uploaded successfully' };
```
Modify message to be something like "transcript successfully parsed"

ZzRanger reviewed (commented):
Looks good overall! Just left a few comments

Commit:
feat(api): Fix hard-coded values in upload.ts and set to return json

Commit:
fix: Added coursecode json file and link to upload.ts

ZzRanger commented on pages/api/transcript/upload.ts:
```diff
@@ -0,0 +1,109 @@
+  try {
+    if (files?.length) {
+      /* Create directory for uploads */
+      const targetPath = path.join(process.cwd(), `/uploads/`);
+      try {
+        await fs.promises.access(targetPath);
+      } catch (e) {
+        await fs.promises.mkdir(targetPath);
+      }
+
+      /* Move uploaded files to directory */
+      for (const file of files) {
+        const tempPath = file[1].filepath;
+        await fs.promises.rename(tempPath, targetPath + file[1].originalFilename);
+      }
+
+      /* Parse the uploaded file */
+      const dataBuffer = fs.readFileSync(
+        process.cwd() + '/uploads/' + files[0][1]['originalFilename'],
+      );
```
Try replacing with 
```
const dataBuffer = Buffer.from(files[0][1]);
```

Commit:
fix: Fixed build error

ZzRanger committed:
chore: Remove intermediary file save step

The pdf file is initially stored in a temp file. Before, the pdf file was copied
to a different location before being read to be parsed. This commit removes that
intermediary file copy step by reading from the temp file itself.

ZzRanger reviewed (approved):
LGTM!
